### PR TITLE
COMPAT: AWS returns InvalidRange error for getObject

### LINKED ISSUE #400: COMPAT: AWS returns InvalidRange error for getObject
AWS return InvalidRange error when sending 'Range' request which specifies a range of bytes which can not be satisfied - because the resource being accessed does not cover this byte range.

**Check against AWS:**
return InvalidRange error _Only if Range start value >=  object size && Range end value >=  object size_

### PR DESCRIPTION
**Pull request by nicolas2bert**
FIXES #400

### CODE REVIEW

File: tests/functional/aws-node-sdk/test/legacy/tests.js [hunk 1]
```diff
@@ -395,6 +395,21 @@ describe('aws-node-sdk test suite as registered user', function testSuite() {
         });
     });
 
+    it('should return InvalidRange if the range the resource being' +
```
Comment (bennettbuchanan) [inline tests/functional/aws-node-sdk/test/legacy/tests.js]: Should be a space after 'being' or before 'accessed'.

File: tests/unit/utils/validateRange.js [hunk 2]
```diff
@@ -3,66 +3,72 @@ import assert from 'assert';
...
+        'format is invalid: start > end', () => {
+        checkRange('bytes=5-4', 10, undefined);
+    });
+
+    it('should return undefined for the range if the range header ' +
+        'format is invalid: start > end', () => {
+        checkRange('bytes=-2-5', 10, undefined);
+    });
+
+    it('should return InvalidRange if the range the resource being' +
```
Comment (bennettbuchanan) [inline tests/unit/utils/validateRange.js]: Should be a space after 'being' or before 'accessed'.

File: tests/functional/aws-node-sdk/test/legacy/tests.js (hunk 1 above)
Comment (nicolas2bert) [inline tests/functional/aws-node-sdk/test/legacy/tests.js]: Thanks

File: tests/unit/utils/validateRange.js (hunk 2 above)
Comment (nicolas2bert) [inline tests/unit/utils/validateRange.js]: Thanks

File: tests/unit/utils/validateRange.js [hunk 3]
```diff
@@ -3,66 +3,71 @@ import assert from 'assert';
...
-        const endGreaterThanStart = 'bytes=5-4';
-        const actualForEndGreater =
-            parseRange(endGreaterThanStart, totalLength);
-        assert.deepStrictEqual(actualForEndGreater, undefined);
-        const negativeStart = 'bytes=-2-5';
-        const actualForNegativeStart =
-            parseRange(negativeStart, totalLength);
-        assert.deepStrictEqual(actualForNegativeStart, undefined);
+        const { range, error } = parseRange(rangeHeader, totalLength);
+        assert.strictEqual(error.code, 416);
```
Comment (bennettbuchanan) [inline tests/unit/utils/validateRange.js]: nice :)

File: lib/api/apiUtils/object/parseRange.js [hunk 4]
```diff
@@ -16,19 +17,19 @@ export function parseRange(rangeHeader, totalLength) {
         || rangeHeader.indexOf('-') < 0
         // Multiple ranges not supported
         || rangeHeader.indexOf(',') > 0) {
-        return range;
+        return { range };
```
Comment (LaurenSpiegel) [inline lib/api/apiUtils/object/parseRange.js]: please update jsdoc

File: tests/functional/aws-node-sdk/test/legacy/tests.js [hunk 5]
```diff
@@ -395,6 +395,21 @@ describe('aws-node-sdk test suite as registered user', function testSuite() {
         });
     });
 
+    it('should return InvalidRange if the range the resource being ' +
```
Comment (rahulreddy) [inline tests/functional/aws-node-sdk/test/legacy/tests.js]: You can reword this as 

> should return InvalidRange if the range of the resource does not cover the byte range

File: tests/unit/utils/validateRange.js [hunk 6]
```diff
@@ -3,66 +3,71 @@ import assert from 'assert';
...
+    });
+
+    it('should return undefined for the range if the range header ' +
+        'format is invalid: value invalid character', () => {
+        checkRange('bytes=%-4', 10, undefined);
+    });
+
+    it('should return undefined for the range if the range header ' +
+    'format is invalid: value not int', () => {
+        checkRange('bytes=4-a', 10, undefined);
```
Comment (rahulreddy) [inline tests/unit/utils/validateRange.js]: Minor, since it's undefined, you can just ignore that instead of specifying it explcitly.

**Comment by nicolas2bert** [comment]
@ironman-machine ok

**Comment by nicolas2bert** [comment]
@ironman-machine ok2

**Comment by nicolas2bert** [comment]
@ironman-machine ok3
